gufe/storage/externalresource.py

```diff
@@ -0,0 +1,161 @@
+import abc
+import hashlib
+import warnings
+import pathlib
+import io
+import contextlib
+import functools
+
+from typing import ClassVar, Union
+
+from gufe.storage.errors import (
+    MissingExternalResourceError, ChangedExternalResourceError
+)
+
+
+class _ForceContext:
+    """Wrapper that forces objects to only be used a context managers.
+
+    Filelike objects can often be used with explicit open/close. This
+    requires the returned filelike to be consumed as a context manager.
+    """
+    def __init__(self, context):
+        self._context = context
+
+    def __enter__(self):
+        return self._context.__enter__()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._context.__exit__(exc_type, exc_value, traceback)
+
+
+class ExternalStorage(abc.ABC):
+    """Abstract base for external storage."""
+
+    allow_changed: ClassVar[bool] = False
+
+    def validate(self, location, metadata):
+        if not self.get_metadata(location) == metadata:
+            msg = (f"Hash mismatch for {location}: this object "
+                   "may have changed.")
+            if not self.allow_changed:
+                # NOTE: having it here instead of in a DataLoader means that
+                # you can only change it for the whole system, instead of
+                # for a specific object
+                raise ChangedExternalResourceError(
+                    msg + " To allow this, set ExternalStorage."
+                    "allow_changed = True"
+                )
+            else:
+                warnings.warn(msg)
+
+    def get_metadata(self, location: str):
+        with self._load_stream(location) as filelike:
+            hasher = hashlib.md5()
+            # TODO: chunking may give better performance
+            hasher.update(filelike.read())
+            digest = hasher.hexdigest()
+
+        return digest
+
+    def get_filename(self, location, metadata) -> str:
+        # we'd like to not need to include the get_filename method, but for
+        # now some consumers of this may not work with filelike
+        self.validate(location, metadata)
+        return self._get_filename(location)
+
+    # @force_context
+    def load_stream(self, location, metadata):
+        self.validate(location, metadata)
+        return _ForceContext(self._load_stream(location))
+
+    @abc.abstractmethod
+    def store(self, location, byte_data):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def exists(self, location):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def delete(self, location):
+        raise NotImplementedError()
```

Comment:
I was tempted to put the docs here, but then annoyingly you'd have to rewrite the docs when you reimplemented this.  Would it be worth making `_delete` the `abc.abstractmethod` to allow the docs for `delete` to be consistent?  Or are individual implementations going to need their own docs?

Reply:
I agree that `_delete` is the better way. Will implement.

This should be seen as an adapter that allows us to abstract the details of storage and provide a limited API for functionality that all backends must support. The only reason not to have `delete` define the API and then call `_delete` would be if the call signature or behavior would depend on the specific subclass. But allowing differences in subclass behavior would break the whole point of having an API with interchangeable backends.